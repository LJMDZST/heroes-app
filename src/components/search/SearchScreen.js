
import React, { useMemo } from 'react';
import queryString from 'query-string'

import { useLocation } from 'react-router-dom';
import { useForm } from '../../hooks/useForm';
import { HeroCard } from '../heroes/HeroCard';
import { getHeroesByName } from '../../selectors/getHeroesByName';

/** Componente que filtra la lista de heroes por nombre */
export const SearchScreen = ({history}) => {
    /** Propiedad del componente que obtiene info relacionada
     *  a la url actual
     */
    const location = useLocation();

    /** queryString, libreria para parsear una query GET */
    const { q = ''} = queryString.parse( location.search );

    /** custom hook useForm() */
    const [values,handleInputChange,reset] = useForm({  
        searchText: q
    });

    /** texto del input de busqueda obtenido por useForm */
    const {searchText} = values 

    /** se llama al selector getHeroesByName usando el hook useMemo para 
     *  que no se haga labusqueda mientras se escribe en el input
     */
    const heroesFiltered = useMemo(() => getHeroesByName( q ), [q] );

    const handleSearch = (e) =>{
        e.preventDefault();

        history.push(`?q=${searchText}`);
        
        reset();
    }

    return (
        <div>
            <h1>Search Hero</h1>
            <hr />

            <div className="row">
                <div className="col-5">
                    <h4>Search Form</h4>
                    <hr />
                    <form onSubmit={handleSearch}>
                        <input 
                            type="text"
                            placeholder="Find your hero.."
                            className="form-control"
                            onChange={handleInputChange}
                            name="searchText"
                            value={searchText}
                        />

                        <button 
                            type="submit"
                            className="btn m-1 btn-block btn-outline-primary"
                           
                        >
                            Search
                        </button>
                    </form>
                </div>

                <div className="col-8">
                    <h4>Results</h4>
                    <hr />
                    {
                        heroesFiltered.map( hero => 
                            <HeroCard 
                                key={hero.id}
                                {...hero}
                            />
                        )
                    }
                </div>


            </div>
        </div>
    )
}
