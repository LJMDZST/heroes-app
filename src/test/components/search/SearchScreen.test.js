import React from 'react';
import {mount} from 'enzyme';
import { MemoryRouter, Route } from 'react-router';
import { SearchScreen } from '../../../components/search/SearchScreen';

describe('Pruebas en <SearchSrcreen />', () => {
    

    test(' debe mostrarse con los valores con defencto', () => {
        const wrapper = mount (
            <MemoryRouter initialEntries={['/search']}>
                <Route path="/search" component={SearchScreen} />
            </MemoryRouter>
        )
        expect(wrapper).toMatchSnapshot();
    })

    test('debe mostrar a Batman y el unput con el valor del queryString', () => {
        const wrapper = mount (
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <Route path="/search" component={SearchScreen} />
            </MemoryRouter>
        )

        expect(wrapper.find('input').prop('value')).toBe('batman');

    })
    
    
})
