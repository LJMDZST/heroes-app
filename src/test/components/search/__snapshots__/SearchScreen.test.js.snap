// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchSrcreen />  debe mostrarse con los valores con defencto 1`] = `
<div>
  <h1>
    Search Hero
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your hero.."
          type="text"
          value=""
        />
        <button
          className="btn m-1 btn-block btn-outline-primary"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-8"
    >
      <h4>
        Results
      </h4>
      <hr />
    </div>
  </div>
</div>
`;
